### ISSUE TITLE
Error when clicking the link in the 'Snapshot Date' column on the 'Running Queries' screen due to the high number of blocks.

### ISSUE THREAD

**Issue by VincenzoMarchese** [trimble-oss/dba-dash#897]
Hello, I encountered the following error when I click on a link in the 'Snapshot Date' column on the 'Running Queries' screen. The row shows a number of 'Blocked queries = 522' and 'Blocked Queries Wait = 269 Days'. The error occurs due to the high number of blocks.

Microsoft.Data.SqlClient.SqlException (0x80131904): **Arithmetic overflow error converting expression to data type int.**
   at Microsoft.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at Microsoft.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at Microsoft.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, SqlCommand command, Boolean callerHasConnectionLock, Boolean asyncClose)
   at Microsoft.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at Microsoft.Data.SqlClient.SqlDataReader.TryHasMoreRows(Boolean& moreRows)
   at Microsoft.Data.SqlClient.SqlDataReader.TryReadInternal(Boolean setTimeout, Boolean& more)
   at Microsoft.Data.SqlClient.SqlDataReader.Read()
   at System.Data.Common.DataAdapter.FillLoadDataRow(SchemaMapping mapping)
   at System.Data.Common.DataAdapter.FillFromReader(DataSet dataset, DataTable datatable, String srcTable, DataReaderContainer dataReader, Int32 startRecord, Int32 maxRecords, DataColumn parentChapterColumn, Object parentChapterValue)
   at System.Data.Common.DataAdapter.Fill(DataTable[] dataTables, IDataReader dataReader, Int32 startRecord, Int32 maxRecords)
   at System.Data.Common.DbDataAdapter.FillInternal(DataSet dataset, DataTable[] datatables, Int32 startRecord, Int32 maxRecords, String srcTable, IDbCommand command, CommandBehavior behavior)
   at System.Data.Common.DbDataAdapter.Fill(DataTable[] dataTables, Int32 startRecord, Int32 maxRecords, IDbCommand command, CommandBehavior behavior)
   at System.Data.Common.DbDataAdapter.Fill(DataTable dataTable)
   at DBADashGUI.Performance.RunningQueries.RunningQueriesSnapshot(DateTime& snapshotDate, Int32 skip) in D:\a\dba-dash\dba-dash\DBADashGUI\Performance\RunningQueries.cs:line 386
   at DBADashGUI.Performance.RunningQueries.LoadSnapshot(DateTime snapshotDate, Int32 skip) in D:\a\dba-dash\dba-dash\DBADashGUI\Performance\RunningQueries.cs:line 400
   at DBADashGUI.Performance.RunningQueries.Dgv_CellContentClick(Object sender, DataGridViewCellEventArgs e) in D:\a\dba-dash\dba-dash\DBADashGUI\Performance\RunningQueries.cs:line 500
   at System.Windows.Forms.DataGridViewCell.OnMouseUpInternal(DataGridViewCellMouseEventArgs e)
   at System.Windows.Forms.DataGridView.OnCellMouseUp(DataGridViewCellMouseEventArgs e)
   at System.Windows.Forms.DataGridView.OnMouseUp(MouseEventArgs e)
   at System.Windows.Forms.Control.WmMouseUp(Message& m, MouseButtons button, Int32 clicks)
   at System.Windows.Forms.Control.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(HWND hWnd, MessageId msg, WPARAM wparam, LPARAM lparam)
ClientConnectionId:3f24d3d4-e4ca-4e34-b89b-b1428c86be48
Error Number:8115,State:2,Class:16

**Comment by DavidWiseman** [comment]
Hi,

I think I've been able to reproduce the same issue by creating some blocking and adjusting the wait times so the cumulative blocking wait exceeds int max value.  I've got a fix for this but it would be good if you could run this on your repository database and let me know if it solves the problem.

```
CREATE OR ALTER FUNCTION dbo.RunningQueriesBlockingRecursiveStats(
	@InstanceID INT,
	@SnapshotDateUTC DATETIME2,
	@session_id SMALLINT
)
RETURNS TABLE 
AS
RETURN
WITH R AS (
	SELECT Q.session_id, Q.wait_time,1 AS IsDirect
	FROM dbo.RunningQueries Q
	WHERE Q.InstanceID = @InstanceID
	AND Q.blocking_session_id=@session_id 
	AND Q.SnapshotDateUTC = @SnapshotDateUTC
	UNION ALL
	SELECT RQ.session_id, RQ.wait_time,0 AS IsDirect
	FROM R 
	JOIN dbo.RunningQueries RQ ON RQ.blocking_session_id = R.session_id
	WHERE RQ.session_id<>@session_id
	AND RQ.InstanceID = @InstanceID
	AND RQ.SnapshotDateUTC = @SnapshotDateUTC
)
SELECT COUNT(*) BlockCountRecursive, 
	ISNULL(SUM(CAST(R.wait_time AS BIGINT)),0) BlockWaitTimeRecursiveMs,
	ISNULL(SUM(R.IsDirect),0) AS BlockCount,
	ISNULL(SUM(CASE WHEN R.IsDirect=1 THEN CAST(R.wait_time AS BIGINT) ELSE 0 END),0) AS BlockWaitTimeMs
FROM R
```

Thanks,

David

**Comment by VincenzoMarchese** [comment]
Hi David, I confirm that the new version of the RunningQueriesBlockingRecursiveStats function solves the problem.
Thank you for producing the fix so quickly!

**Comment by DavidWiseman** [comment]
The issue came in at a good time.  Thanks for confirming that it resolved the issue.  It will be included in the next release.

**Comment by DavidWiseman** [comment]
Fixed in 3.5.2 🎉
